src/extensions.rs

```diff
@@ -31,7 +30,7 @@ impl Hasher for IdHasher {
 ///
 /// `Extensions` can be used by `Request` and `Response` to store
 /// extra data derived from the underlying protocol.
-#[derive(Default)]
+#[derive(Default, Clone)]
```

Comment:
The `derive(Clone)` here means that the `NotCloneExtension` will remain in the new `AnyMap`, as `NotCloneExtension(None)`. Is this what we want?

An alternative here might be:

- The `HashMap` store values as `enum Value { Clone(Box<dyn CloneAny>), NotClone(Box<dyn Any>) }`,
- `Clone` filters out the `Value::NotClone`,
- Have two methods `fn insert` and `fn insert_clone`, only the latter requiring `Clone`.

Reply:
> The derive(Clone) here means that the NotCloneExtension will remain in the new AnyMap, as NotCloneExtension(None). Is this what we want?

Im not totally following what you mean by this?

Reply:
```rust
let mut extensions = Extensions::new();
extensions.insert(NotCloneExtension::new(3_u32));
let extensions_b = extensions.clone();
let value = extensions_b.get::<NotCloneExtension<u32>>();
```

`value` here is `Some(NotCloneExtension(None))` from what I read from the implementation. Is this what we want? If we take the changes described above and provide a hand written `Clone` implementation we could filter out the non-`Clone` extensions.

Having two separate methods `fn insert` and `fn insert_clone` then storing the `enum Value`, has the additional benefit of:

- It's an addition rather than a breaking change.
- It doesn't require a newtype when using `get` - e.g. `get::<T>` will retrieve `T` when it's inserted via `insert` or `insert_clone`. This is in contrast to `get::<NotCloneExtension<T>>` and `get::<T>`.

Reply:
Something like this is what I was imagining

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=318f42c9a2477340c76ddf52aeefd908

It doesn't compile yet, but I think it would if we employed some more of the `anymap` tricks?